### DISCUSSION TITLE
PowerShell Script to Force Downloads

### DISCUSSION THREAD

**Discussion by jasonpatrickellykrause** [RandomNinjaAtk/arr-scripts#147] [Show and tell]
I move from Windows Lidarr to Docker Lidarr last week so I could use [arr-scripts](https://github.com/RandomNinjaAtk/arr-scripts/blob/main/lidarr/readme.md) to pull from Tidal and Deezer. It was a great learning experience for sure. 

The aar-script for Lidarr did a great job of downloading any missing music it could find. The only thing I could not figure out how to do was make the scripts upgrade albums I already downloaded. To address this, I cobbled together a couple of PowerShell scripts to make the magic happen.

## Step One

Rename your mp3 files to mp3old.

```PowerShell
$musicFolder = "D:\Music"

# Get all directories in Music folder
# Consider adding the -Filter option to perform the rename in small batches.
# -Filter a* would return all folders that start with the letter A.
$folders = Get-Childitem $musicFolder -Directory

foreach ($folder in $folders) {
    Get-ChildItem -Path $folder -Recurse -Filter "*.mp3" | Rename-Item -NewName { [io.path]::ChangeExtension($_.name, "mp3old") }
}
```

## Step Two

Refresh artists and wait for scripts to complete search. This will take hours or days depending on how many files you renamed.

## Step Three

Search for any folders that now contain mp3old and flac. If there is flac, delete the mp3olds. If there is no flac, rename the mp3old back. 

```PowerShell
$musicFolder = "D:\Music"

# Get all Directory ojbects in Music folder
$folders = Get-Childitem $musicFolder  -Directory -Recurse

foreach ($folder in $folders) {

    # Count mp3old files in folder
    $mp3OldCount = (Get-ChildItem -Path $folder -Filter "*.mp3old" | Measure-Object).Count
    
    # If mpp3old count is greater than zero, proceed to check for flac files
    if ($mp3OldCount -gt 0) {

        # Count flac files in folder
        $flacCount = (Get-ChildItem -Path $folder -Filter "*.flac" | Measure-Object).Count
c
        # If flac count is greater than zero, remove mp3old files
        if ($flacCount -gt 0) {
            Get-ChildItem -Path $folder -Recurse -Filter "*.mp3old" | Remove-Item -WhatIf
        }
        # If no flac files found, rename mp3old back to mp3
        else {
            Get-ChildItem -Path $folder -Recurse -Filter "*.mp3old" | Rename-Item -NewName { [io.path]::ChangeExtension($_.name, "mp3") } -WhatIf
        } 
    }    
}
```

**Comment by jasonpatrickellykrause** [comment]
I have not tested this for every use case, which is why I left the `-WhatIf` in the script itself.

**Reply by RandomNinjaAtk** [reply]
It should upgrade stuff you have, without the need of the script… that assumes your media is still wanted and has not met the cutoff… which means you have to configure lidarr correctly for it to work.

**Reply by jasonpatrickellykrause** [reply]
Dang!

Let me look at Lidarr and see where I went wrong. I assume I lost something when I moved to Docker.

**Reply by jasonpatrickellykrause** [reply]
OK, I figured it out, I missed this checkmark when I created my quality profile. 
![image](https://github.com/RandomNinjaAtk/arr-scripts/assets/8408603/e854bb8c-91ba-43cd-a5fc-d86949d4c84d)
